### PR TITLE
[CCR] Validate cluster alias in follow APIs

### PR DESCRIPTION
**Pull request by martijnvg**
Spawned from #33999

### CODE REVIEW

**Review by javanna — COMMENTED** [review]
left a comment

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportCreateAndFollowIndexAction.java [hunk 1]
```diff
@@ -102,6 +104,7 @@ protected void masterOperation(
             return;
         }
         final String[] indices = new String[]{request.getFollowRequest().getLeaderIndex()};
+        validateClusterAlias(remoteClusterService.getRemoteClusterNames(), request.getFollowRequest().getLeaderIndex());
         final Map<String, List<String>> remoteClusterIndices = remoteClusterService.groupClusterIndices(indices, s -> false);
```
[the remark this answers is not in the document]
Comment (martijnvg) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportCreateAndFollowIndexAction.java]: Yes, I can do that. Sorry I think I misunderstood what we discussed yesterday. I will change this.

[the remark this answers is not in the document]
Comment (martijnvg) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportCreateAndFollowIndexAction.java]: Thanks @javanna, I pushed: 965ded77ed9cd319fa0dd23932b17b474057cf6b

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportFollowIndexAction.java [hunk 2]
```diff
@@ -297,6 +293,19 @@ static void validate(
         followerMapperService.merge(leaderIndex, MapperService.MergeReason.MAPPING_RECOVERY);
     }
 
+    static Tuple<String, String> splitClusterAliasAndIndexName(Set<String> remoteClusterNames, String leaderIndexExpression) {
+        int indexOf = leaderIndexExpression.indexOf(RemoteClusterService.REMOTE_CLUSTER_INDEX_SEPARATOR);
+        if (indexOf != -1) {
+            String clusterAlias = leaderIndexExpression.substring(0, indexOf);
+            if (remoteClusterNames.contains(clusterAlias) == false) {
+                throw new IllegalArgumentException("unknown cluster alias [" + clusterAlias + "]");
```
Comment (javanna) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportFollowIndexAction.java]: don't you need to handle the case where you have a local index that contains ":" ?

Comment (martijnvg) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportFollowIndexAction.java]: Yes, but that will be addressed in a different PR. This change is just about validating whether the cluster alias exists and whether it has been configured.

That validation can be added to `FollowIndexAction.Request.validate()` method. If a leaderIndex contains more than 1 `:` then we can fail the request.

Comment (javanna) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportFollowIndexAction.java]: but is the plan to go local if no remote cluster is found and there is an index with that name (although it contains ":") ? Seems like we may end up duplicating the bw comp logic from RemoteClusterService, and we already duplicate the splitting logic. I am fine with changing this later but I am not 100% convinced that this new method belongs here.

**Review by jasontedor — CHANGES_REQUESTED** [review]
Let's separate the `leader_index` request body parameters into `cluster_alias` and `leader_index` instead of specifying as `foo:bar` a user would say `"cluster_alias": "foo", "leader_index": "bar"`.

**Comment by martijnvg** [comment]
Cluster alias will be split of from the leader index field into a separate field and that will change how cluster aliases will be validated.

**Pull request closed without merging**